File: app/src/main/java/c/m/storyapp/login/data/repository/LoginRepositoryImpl.kt
package c.m.storyapp.login.data.repository

import c.m.storyapp.R
import c.m.storyapp.common.util.Resource
import c.m.storyapp.common.util.SimpleResource
import c.m.storyapp.common.util.UIText
import c.m.storyapp.data_store.domain.repository.DataStorePreferencesRepository
import c.m.storyapp.login.data.mapper.toLoginResponse
import c.m.storyapp.login.data.remote.LoginAPI
import c.m.storyapp.login.domain.model.LoginResponse
import c.m.storyapp.login.domain.repository.LoginRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import retrofit2.HttpException
import timber.log.Timber
import java.io.IOException
import java.net.UnknownHostException
import javax.inject.Inject

class LoginRepositoryImpl @Inject constructor(
    private val loginAPI: LoginAPI
) : LoginRepository {
    override suspend fun postLogin(email: String, password: String): Flow<Resource<LoginResponse>> {
        return flow {
            // Loading State
            emit(Resource.Loading(isLoading = true))

            try {
                // Loading State
                emit(Resource.Loading(isLoading = false))

                val postLogin = loginAPI.postLogin(email, password)

                if (postLogin.error) {
                    // Error State
                    emit(Resource.Error(message = UIText.DynamicString(postLogin.message)))
                }

                // Success State
                emit(Resource.Success(data = postLogin.toLoginResponse()))
            } catch (e: HttpException) {
                // Loading State
                emit(Resource.Loading(isLoading = false))

                // Error State
                Timber.e(e.localizedMessage)
                emit(
                    Resource.Error(message = UIText.StringResource(R.string.error_internet_problem))
                )
            } catch (e: IOException) {
                // Loading State
                emit(Resource.Loading(isLoading = false))

                // Error State
                Timber.e(e.localizedMessage)
                emit(
                    Resource.Error(message = UIText.StringResource(R.string.error_internet_problem))
                )
            } catch (e: UnknownHostException) {
                // Loading State
                emit(Resource.Loading(isLoading = false))

                // Error State
                Timber.e(e.localizedMessage)
                emit(
                    Resource.Error(message = UIText.StringResource(R.string.error_unknown))
                )
            }
        }.flowOn(Dispatchers.IO)
    }
}

File: lib/login/data/repository/login_repository_impl.dart
import 'dart:async';

import 'package:injectable/injectable.dart';
import 'package:story_app/common/data/remote/api_services.dart';
import 'package:story_app/common/util/resource.dart';
import 'package:story_app/login/data/mapper/login_mapper.dart';
import 'package:story_app/login/data/remote/dto/login_response_dto.dart';
import 'package:story_app/login/domain/model/login_response.dart';
import 'package:story_app/login/domain/repository/login_repository.dart';

@Injectable(as: LoginRepository)
class LoginRepositoryImpl implements LoginRepository {
  final ApiServices _apiServices;

  LoginRepositoryImpl(this._apiServices);

  @override
  Future<Resource<LoginResponse>> postLogin(
      String email, String password) async {
    Resource<LoginResponse> result;

    try {
      final responseLogin = await _apiServices.postLogin(email, password);

      if (responseLogin.isSuccessful) {
        result = Resource.success(
          LoginResponseDTO.fromJson(responseLogin.body).toLoginResponse(),
        );
      } else {
        result = Resource.error(
          LoginResponseDTO.fromJson(responseLogin.error as Map<String, Object?>)
              .toLoginResponse()
              .message,
          LoginResponseDTO.fromJson(responseLogin.error as Map<String, Object?>)
              .toLoginResponse(),
        );
      }
    } catch (e) {
      result = Resource.error(e.toString(), null);
    } finally {
      _apiServices.client.dispose();
    }

    return result;
  }
}
